File: scripts/hybrid_vision.py
"""
Hybrid ONNX + Ollama Vision System
Uses ONNX for fast GPU-accelerated image preprocessing and validation,
then sends optimized images to Ollama for vision-language understanding.
"""

import onnxruntime as ort
import numpy as np
import requests
import base64
import time
import json
from pathlib import Path
from PIL import Image
from io import BytesIO
from typing import Optional, Tuple


class HybridVisionSystem:
    """
    Combines ONNX preprocessing with Ollama vision model.
    - ONNX: Fast GPU preprocessing, validation, feature extraction
    - Ollama: Vision-language understanding and text generation
    """
    
    def __init__(
        self,
        onnx_model_path: str = "D:/G.R.I.M/data/models/vision/llama3.2-vision-onnx/vision_encoder.onnx",
        ollama_model: str = "llama3.2-vision:11b",
        ollama_url: str = "http://localhost:11434"
    ):
        self.ollama_model = ollama_model
        self.ollama_url = ollama_url
        self.onnx_session = None
        
        # Load ONNX model for preprocessing
        model_path = Path(onnx_model_path)
        if model_path.exists():
            print(f"Loading ONNX preprocessor from {model_path}...")
            self._load_onnx_model(model_path)
        else:
            print(f"⚠️  ONNX model not found at {model_path}")
            print("   Continuing without ONNX acceleration")
    
    def _load_onnx_model(self, model_path: Path):
        """Load ONNX model for GPU-accelerated preprocessing."""
        try:
            sess_options = ort.SessionOptions()
            sess_options.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_ALL
            sess_options.intra_op_num_threads = 4
            
            # Prefer CUDA for GPU acceleration
            providers = []
            if 'CUDAExecutionProvider' in ort.get_available_providers():
                print("✅ Using GPU acceleration (CUDA)")
                providers.append('CUDAExecutionProvider')
            else:
                print("⚠️  Using CPU for preprocessing")
            providers.append('CPUExecutionProvider')
            
            self.onnx_session = ort.InferenceSession(
                str(model_path),
                sess_options,
                providers=providers
            )
            print(f"✅ ONNX preprocessor loaded")
            
        except Exception as e:
            print(f"⚠️  Failed to load ONNX model: {e}")
            self.onnx_session = None
    
    def preprocess_image_onnx(self, image: Image.Image) -> Optional[np.ndarray]:
        """
        Use ONNX for fast GPU-accelerated image preprocessing.
        Returns visual embeddings or None if ONNX not available.
        """
        if not self.onnx_session:
            return None
        
        try:
            # Resize and normalize image for ONNX model
            # llama3.2-vision expects 560x560 images
            img_resized = image.resize((560, 560), Image.Resampling.LANCZOS)
            img_array = np.array(img_resized).astype(np.float32) / 255.0
            
            # Convert HWC to CHW format
            img_array = np.transpose(img_array, (2, 0, 1))
            
            # Normalize with ImageNet stats
            mean = np.array([0.485, 0.456, 0.406]).reshape(3, 1, 1)
            std = np.array([0.229, 0.224, 0.225]).reshape(3, 1, 1)
            img_array = (img_array - mean) / std
            
            # Add batch dimension
            img_array = np.expand_dims(img_array, axis=0)
            
            # Run ONNX inference
            start_time = time.time()
            inputs = {self.onnx_session.get_inputs()[0].name: img_array}
            outputs = self.onnx_session.run(None, inputs)
            preprocess_time = (time.time() - start_time) * 1000
            
            print(f"  ONNX preprocessing: {preprocess_time:.1f}ms")
            print(f"  Embeddings shape: {outputs[0].shape}")
            
            return outputs[0]
            
        except Exception as e:
            print(f"⚠️  ONNX preprocessing failed: {e}")
            return None
    
    def optimize_image_for_ollama(self, image: Image.Image) -> Tuple[str, int]:
        """
        Optimize image for Ollama API (resize, compress, encode).
        Returns base64 string and size in bytes.
        """
        # Resize to reasonable size (Ollama handles resizing, but smaller = faster)
        max_size = 896  # llama3.2-vision can handle up to 896x896
        
        # Calculate resize maintaining aspect ratio
        width, height = image.size
        if width > max_size or height > max_size:
            ratio = min(max_size / width, max_size / height)
            new_size = (int(width * ratio), int(height * ratio))
            image = image.resize(new_size, Image.Resampling.LANCZOS)
        
        # Convert to RGB if needed
        if image.mode != 'RGB':
            image = image.convert('RGB')
        
        # Compress to JPEG with good quality
        buffer = BytesIO()
        image.save(buffer, format='JPEG', quality=85, optimize=True)
        img_bytes = buffer.getvalue()
        
        # Encode to base64
        img_b64 = base64.b64encode(img_bytes).decode('utf-8')
        
        return img_b64, len(img_bytes)
    
    def query_ollama_vision(
        self,
        image: Image.Image,
        prompt: str,
        use_onnx_preprocess: bool = True
    ) -> dict:
        """
        Query Ollama vision model with optional ONNX preprocessing.
        
        Args:
            image: PIL Image to analyze
            prompt: Text prompt for the vision model
            use_onnx_preprocess: Whether to use ONNX for preprocessing validation
        
        Returns:
            Dictionary with response and timing information
        """
        print(f"\n{'='*70}")
        print(f"Querying Vision Model: {self.ollama_model}")
        print(f"{'='*70}")
        print(f"Prompt: {prompt}")
        print(f"{'-'*70}")
        
        result = {
            'prompt': prompt,
            'response': None,
            'error': None,
            'timings': {}
        }
        
        # Optional ONNX preprocessing for validation/speed check
        if use_onnx_preprocess and self.onnx_session:
            print("\n1️⃣  ONNX Preprocessing...")
            embeddings = self.preprocess_image_onnx(image)
            if embeddings is not None:
                print(f"   ✅ Image validated and preprocessed on GPU")
                result['timings']['onnx_preprocess_ms'] = embeddings
        
        # Optimize image for Ollama
        print("\n2️⃣  Optimizing image for Ollama...")
        start_time = time.time()
        img_b64, img_size = self.optimize_image_for_ollama(image)
        optimize_time = (time.time() - start_time) * 1000
        
        print(f"   Original size: {image.size}")
        print(f"   Optimized size: {img_size / 1024:.1f} KB")
        print(f"   Optimization time: {optimize_time:.1f}ms")
        
        result['timings']['image_optimize_ms'] = optimize_time
        result['timings']['image_size_kb'] = img_size / 1024
        
        # Query Ollama API
        print("\n Sending to Ollama...")
        try:
            payload = {
                'model': self.ollama_model,
                'prompt': prompt,
                'images': [img_b64],
                'stream': False
            }
            
            start_time = time.time()
            response = requests.post(
                f"{self.ollama_url}/api/generate",
                json=payload,
                timeout=60
            )
            api_time = (time.time() - start_time) * 1000
            
            result['timings']['ollama_api_ms'] = api_time
            
            if response.status_code == 200:
                data = response.json()
                result['response'] = data.get('response', '')
                
                # Extract additional timing info from Ollama
                if 'total_duration' in data:
                    result['timings']['ollama_total_ns'] = data['total_duration']
                    result['timings']['ollama_total_ms'] = data['total_duration'] / 1e6
                
                if 'load_duration' in data:
                    result['timings']['ollama_load_ms'] = data['load_duration'] / 1e6
                
                if 'prompt_eval_duration' in data:
                    result['timings']['ollama_prompt_eval_ms'] = data['prompt_eval_duration'] / 1e6
                
                if 'eval_duration' in data:
                    result['timings']['ollama_eval_ms'] = data['eval_duration'] / 1e6
                
                print(f"   ✅ Response received in {api_time:.0f}ms")
                
            else:
                error_msg = f"Ollama API error: {response.status_code}"
                result['error'] = error_msg
                print(f"   ❌ {error_msg}")
                
        except requests.Timeout:
            result['error'] = "Request timed out after 60s"
            print(f"   ❌ {result['error']}")
        except Exception as e:
            result['error'] = str(e)
            print(f"   ❌ Error: {e}")
        
        return result
    
    def print_result(self, result: dict):
        """Pretty print the vision query result."""
        print(f"\n{'='*70}")
        print("RESULT")
        print(f"{'='*70}")
        
        if result['error']:
            print(f"❌ Error: {result['error']}")
        elif result['response']:
            print(f"✅ Response:")
            print(f"\n{result['response']}\n")
        
        # Print timing breakdown
        if result['timings']:
            print(f"{'-'*70}")
            print("Timing Breakdown:")
            print(f"{'-'*70}")
            
            timings = result['timings']
            
            if 'onnx_preprocess_ms' in timings:
                print(f"  ONNX preprocessing: {timings['onnx_preprocess_ms']:.1f}ms")
            
            if 'image_optimize_ms' in timings:
                print(f"  Image optimization: {timings['image_optimize_ms']:.1f}ms")
                print(f"  Optimized image size: {timings['image_size_kb']:.1f}KB")
            
            if 'ollama_load_ms' in timings:
                print(f"  Ollama model load: {timings['ollama_load_ms']:.1f}ms")
            
            if 'ollama_prompt_eval_ms' in timings:
                print(f"  Ollama prompt eval: {timings['ollama_prompt_eval_ms']:.1f}ms")
            
            if 'ollama_eval_ms' in timings:
                print(f"  Ollama generation: {timings['ollama_eval_ms']:.1f}ms")
            
            if 'ollama_total_ms' in timings:
                print(f"  Ollama total: {timings['ollama_total_ms']:.0f}ms")
            
            if 'ollama_api_ms' in timings:
                print(f"  API round-trip: {timings['ollama_api_ms']:.0f}ms")
        
        print(f"{'='*70}\n")


def test_hybrid_system():
    """Test the hybrid ONNX + Ollama vision system."""
    
    print("Initializing Hybrid Vision System...")
    system = HybridVisionSystem()
    
    # Test with a sample image
    print("\nCreating test image...")
    test_image = Image.new('RGB', (800, 600), color=(73, 109, 137))
    
    # Add some text to the image for testing
    from PIL import ImageDraw, ImageFont
    draw = ImageDraw.Draw(test_image)
    
    try:
        font = ImageFont.truetype("arial.ttf", 60)
    except:
        font = ImageFont.load_default()
    
    draw.text((50, 250), "Hello from GRIM!", fill=(255, 255, 255), font=font)
    
    # Test queries
    test_queries = [
        "What do you see in this image?",
        "Describe the text and colors in detail.",
        "What is written in this image?"
    ]
    
    for i, query in enumerate(test_queries, 1):
        print(f"\n{'#'*70}")
        print(f"Test Query {i}/{len(test_queries)}")
        print(f"{'#'*70}")
        
        result = system.query_ollama_vision(
            test_image,
            query,
            use_onnx_preprocess=True
        )
        
        system.print_result(result)
        
        if i < len(test_queries):
            print("\nWaiting 2s before next query...")
            time.sleep(2)
    
    print(f"\n{'='*70}")
    print("✅ Hybrid Vision System Test Complete")
    print(f"{'='*70}")


def interactive_mode():
    """Interactive mode for testing with custom images and prompts."""
    
    print("=" * 70)
    print("Hybrid Vision System - Interactive Mode")
    print("=" * 70)
    
    system = HybridVisionSystem()
    
    while True:
        print("\nOptions:")
        print("  1. Load image from file")
        print("  2. Use test image")
        print("  3. Exit")
        
        choice = input("\nChoice: ").strip()
        
        if choice == '3':
            print("Goodbye!")
            break
        
        # Get image
        image = None
        if choice == '1':
            path = input("Image path: ").strip()
            try:
                image = Image.open(path)
                print(f"✅ Loaded {image.size[0]}x{image.size[1]} image")
            except Exception as e:
                print(f"❌ Failed to load image: {e}")
                continue
        elif choice == '2':
            image = Image.new('RGB', (800, 600), color=(73, 109, 137))
            from PIL import ImageDraw
            draw = ImageDraw.Draw(image)
            draw.text((50, 250), "Test Image", fill=(255, 255, 255))
            print("✅ Created test image")
        else:
            print("Invalid choice")
            continue
        
        # Get prompt
        prompt = input("\nPrompt: ").strip()
        if not prompt:
            prompt = "What do you see in this image?"
        
        # Query
        result = system.query_ollama_vision(image, prompt)
        system.print_result(result)


if __name__ == "__main__":
    import sys
    
    if len(sys.argv) > 1 and sys.argv[1] == '--interactive':
        interactive_mode()
    else:
        test_hybrid_system()

File: vision/hybrid_vision.cpp
/**
 * @file hybrid_vision.cpp
 * @brief Implementation of Hybrid ONNX + Ollama Vision System
 */

#include "hybrid_vision.hpp"
#include <iostream>
#include <fstream>
#include <sstream>
#include <algorithm>
#include <cmath>
#include <cpr/cpr.h>
#include <nlohmann/json.hpp>

// STB image library (implementation is in popup_ui/stb_image_impl.cpp)
#include <stb/stb_image.h>

// Use OpenCV for image processing (resize, JPEG encoding)
#include <opencv2/opencv.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>

namespace grim {
namespace vision {

// Base64 encoding helper
static std::string base64_encode(const std::vector<uint8_t>& data) {
    static const char* base64_chars = 
        "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        "abcdefghijklmnopqrstuvwxyz"
        "0123456789+/";
    
    std::string result;
    result.reserve(((data.size() + 2) / 3) * 4);
    
    size_t i = 0;
    while (i < data.size()) {
        uint32_t octet_a = i < data.size() ? data[i++] : 0;
        uint32_t octet_b = i < data.size() ? data[i++] : 0;
        uint32_t octet_c = i < data.size() ? data[i++] : 0;
        
        uint32_t triple = (octet_a << 16) | (octet_b << 8) | octet_c;
        
        result.push_back(base64_chars[(triple >> 18) & 0x3F]);
        result.push_back(base64_chars[(triple >> 12) & 0x3F]);
        result.push_back(base64_chars[(triple >> 6) & 0x3F]);
        result.push_back(base64_chars[triple & 0x3F]);
    }
    
    // Add padding
    size_t padding = (3 - (data.size() % 3)) % 3;
    for (size_t p = 0; p < padding; p++) {
        result[result.size() - 1 - p] = '=';
    }
    
    return result;
}

// ============================================================================
// HybridVisionSystem Implementation
// ============================================================================

HybridVisionSystem::HybridVisionSystem(const VisionConfig& config)
    : config_(config)
{
    std::cout << "Initializing Hybrid Vision System..." << std::endl;
    
    // Initialize ONNX Runtime
    if (config_.use_onnx_preprocessing) {
        if (init_onnx()) {
            std::cout << "✅ ONNX preprocessing enabled" << std::endl;
        } else {
            std::cout << "⚠️  ONNX preprocessing unavailable, continuing without it" << std::endl;
        }
    }
}

HybridVisionSystem::~HybridVisionSystem() {
    // Cleanup handled by smart pointers
}

bool HybridVisionSystem::init_onnx() {
#ifdef GRIM_HAS_ONNXRUNTIME
    try {
        onnx_env_ = std::make_unique<Ort::Env>(ORT_LOGGING_LEVEL_WARNING, "HybridVision");
        
        // Create session options
        onnx_session_options_ = std::make_unique<Ort::SessionOptions>();
        onnx_session_options_->SetIntraOpNumThreads(config_.onnx_threads);
        onnx_session_options_->SetGraphOptimizationLevel(
            GraphOptimizationLevel::ORT_ENABLE_ALL
        );
        
        // Try to use CUDA if available and requested
        if (config_.use_gpu) {
            try {
                OrtCUDAProviderOptions cuda_options;
                cuda_options.device_id = 0;
                onnx_session_options_->AppendExecutionProvider_CUDA(cuda_options);
                using_cuda_ = true;
                std::cout << "✅ Using GPU acceleration (CUDA)" << std::endl;
            } catch (const std::exception& e) {
                std::cout << "⚠️  CUDA not available: " << e.what() << std::endl;
                std::cout << "   Falling back to CPU" << std::endl;
            }
        }
        
        // Load the model
#ifdef _WIN32
        std::wstring model_path_w(config_.onnx_model_path.begin(), 
                                  config_.onnx_model_path.end());
        onnx_session_ = std::make_unique<Ort::Session>(
            *onnx_env_,
            model_path_w.c_str(),
            *onnx_session_options_
        );
#else
        onnx_session_ = std::make_unique<Ort::Session>(
            *onnx_env_,
            config_.onnx_model_path.c_str(),
            *onnx_session_options_
        );
#endif
        
        // Get input/output names
        Ort::AllocatorWithDefaultOptions allocator;
        
        size_t num_inputs = onnx_session_->GetInputCount();
        for (size_t i = 0; i < num_inputs; i++) {
            auto name = onnx_session_->GetInputNameAllocated(i, allocator);
            input_names_.push_back(name.get());
        }
        
        size_t num_outputs = onnx_session_->GetOutputCount();
        for (size_t i = 0; i < num_outputs; i++) {
            auto name = onnx_session_->GetOutputNameAllocated(i, allocator);
            output_names_.push_back(name.get());
        }
        
        std::cout << "✅ ONNX model loaded from: " << config_.onnx_model_path << std::endl;
        return true;
        
    } catch (const std::exception& e) {
        std::cerr << "Failed to initialize ONNX: " << e.what() << std::endl;
#ifdef GRIM_HAS_ONNXRUNTIME
        onnx_session_.reset();
#endif
        return false;
    }
#else
    return false;
#endif
}

float HybridVisionSystem::preprocess_with_onnx(
    const ImageData& image,
    std::vector<float>& out_embeddings
) {
#ifdef GRIM_HAS_ONNXRUNTIME
    if (!onnx_session_) {
        return 0.0f;
    }
    
    auto start = std::chrono::high_resolution_clock::now();
    
    try {
        // Resize to 560x560 (llama3.2-vision expected size)
        ImageData resized = image;
        resize_image(resized, 560);
        
        // Normalize and convert to CHW format
        std::vector<float> input_tensor_values(1 * 3 * 560 * 560);
        
        const float mean[3] = {0.485f, 0.456f, 0.406f};
        const float std[3] = {0.229f, 0.224f, 0.225f};
        
        for (int c = 0; c < 3; c++) {
            for (int h = 0; h < 560; h++) {
                for (int w = 0; w < 560; w++) {
                    int pixel_idx = (h * 560 + w) * 3 + c;
                    int tensor_idx = c * 560 * 560 + h * 560 + w;
                    
                    float pixel = resized.pixels[pixel_idx] / 255.0f;
                    input_tensor_values[tensor_idx] = (pixel - mean[c]) / std[c];
                }
            }
        }
        
        // Create input tensor
        std::vector<int64_t> input_shape = {1, 3, 560, 560};
        auto memory_info = Ort::MemoryInfo::CreateCpu(
            OrtArenaAllocator, OrtMemTypeDefault
        );
        
        Ort::Value input_tensor = Ort::Value::CreateTensor<float>(
            memory_info,
            input_tensor_values.data(),
            input_tensor_values.size(),
            input_shape.data(),
            input_shape.size()
        );
        
        // Prepare input/output
        std::vector<const char*> input_names_cstr;
        for (const auto& name : input_names_) {
            input_names_cstr.push_back(name.c_str());
        }
        
        std::vector<const char*> output_names_cstr;
        for (const auto& name : output_names_) {
            output_names_cstr.push_back(name.c_str());
        }
        
        // Run inference
        auto output_tensors = onnx_session_->Run(
            Ort::RunOptions{nullptr},
            input_names_cstr.data(),
            &input_tensor,
            1,
            output_names_cstr.data(),
            output_names_cstr.size()
        );
        
        // Extract embeddings
        float* output_data = output_tensors[0].GetTensorMutableData<float>();
        auto type_info = output_tensors[0].GetTensorTypeAndShapeInfo();
        size_t total_elements = type_info.GetElementCount();
        
        out_embeddings.assign(output_data, output_data + total_elements);
        
        auto end = std::chrono::high_resolution_clock::now();
        float duration = std::chrono::duration<float, std::milli>(end - start).count();
        
        return duration;
        
    } catch (const std::exception& e) {
        std::cerr << "ONNX preprocessing error: " << e.what() << std::endl;
        return 0.0f;
    }
#else
    (void)image; (void)out_embeddings;
    return 0.0f;
#endif
}

void HybridVisionSystem::resize_image(ImageData& image, int max_size) {
    if (image.width <= max_size && image.height <= max_size) {
        return;
    }
    
    // Calculate new dimensions maintaining aspect ratio
    float ratio = std::min(
        static_cast<float>(max_size) / image.width,
        static_cast<float>(max_size) / image.height
    );
    
    int new_width = static_cast<int>(image.width * ratio);
    int new_height = static_cast<int>(image.height * ratio);
    
    // Convert to OpenCV Mat
    cv::Mat src(image.height, image.width, CV_8UC3, image.pixels.data());
    cv::Mat dst;
    
    // Resize using OpenCV (high quality Lanczos interpolation)
    cv::resize(src, dst, cv::Size(new_width, new_height), 0, 0, cv::INTER_LANCZOS4);
    
    // Copy back to ImageData
    image.width = new_width;
    image.height = new_height;
    image.pixels.assign(dst.data, dst.data + (new_width * new_height * image.channels));
}

float HybridVisionSystem::optimize_for_ollama(
    const ImageData& image,
    std::string& out_base64
) {
    auto start = std::chrono::high_resolution_clock::now();
    
    // Resize if needed
    ImageData optimized = image;
    resize_image(optimized, config_.max_image_size);
    
    // Convert to OpenCV Mat
    cv::Mat img_mat(optimized.height, optimized.width, CV_8UC3, optimized.pixels.data());
    
    // Encode to JPEG using OpenCV
    std::vector<uint8_t> jpeg_buffer;
    std::vector<int> params = {cv::IMWRITE_JPEG_QUALITY, config_.jpeg_quality};
    
    if (!cv::imencode(".jpg", img_mat, jpeg_buffer, params)) {
        std::cerr << "Failed to encode JPEG" << std::endl;
        return 0.0f;
    }
    
    // Encode to base64
    out_base64 = base64_encode(jpeg_buffer);
    
    auto end = std::chrono::high_resolution_clock::now();
    return std::chrono::duration<float, std::milli>(end - start).count();
}

bool HybridVisionSystem::query_ollama(
    const std::string& prompt,
    const std::string& image_b64,
    VisionResult& result
) {
    auto start = std::chrono::high_resolution_clock::now();
    
    try {
        std::string url = config_.ollama_url + "/api/generate";
        
        std::cout << "   Ollama URL: " << url << std::endl;
        std::cout << "   Model: " << config_.ollama_model << std::endl;
        std::cout << "   Image size: " << (image_b64.size() / 1024) << " KB (base64)" << std::endl;
        std::cout << "   Timeout: " << config_.timeout_seconds << "s" << std::endl;
        
        // Build JSON payload
        nlohmann::json payload = {
            {"model", config_.ollama_model},
            {"prompt", prompt},
            {"images", {image_b64}},
            {"stream", false}
        };
        
        std::string payload_str = payload.dump();
        std::cout << "   Payload size: " << (payload_str.size() / 1024) << " KB" << std::endl;
        
        std::cout << "   Sending POST request..." << std::endl;
        
        // Send POST request using CPR with appropriate timeout for vision models
        // Vision models can take 20-60 seconds to process, so we need generous timeouts
        cpr::Response response = cpr::Post(
            cpr::Url{url},
            cpr::Header{{"Content-Type", "application/json"}},
            cpr::Body{payload_str},
            cpr::Timeout{std::chrono::milliseconds(config_.timeout_seconds * 1000)},
            cpr::ConnectTimeout{std::chrono::milliseconds(10000)}  // 10 second connect timeout
            // Note: Removed LowSpeed limit - vision models process slowly but are working
        );
        
        auto end = std::chrono::high_resolution_clock::now();
        result.timings.api_roundtrip_ms = 
            std::chrono::duration<float, std::milli>(end - start).count();
        
        std::cout << "   Response received in " << result.timings.api_roundtrip_ms << "ms" << std::endl;
        
        // Check for errors
        if (response.error.code != cpr::ErrorCode::OK) {
            result.error = "Connection error: " + response.error.message;
            std::cout << "   ❌ " << result.error << std::endl;
            return false;
        }
        
        std::cout << "   HTTP Status: " << response.status_code << std::endl;
        
        if (response.status_code != 200) {
            result.error = "HTTP " + std::to_string(response.status_code);
            if (!response.text.empty()) {
                result.error += ": " + response.text.substr(0, 200);  // First 200 chars
            }
            std::cout << "   ❌ " << result.error << std::endl;
            return false;
        }
        
        // Parse response
        std::cout << "   Parsing JSON response..." << std::endl;
        auto response_json = nlohmann::json::parse(response.text);
        
        result.response = response_json.value("response", "");
        
        std::cout << "   Response length: " << result.response.size() << " chars" << std::endl;
        
        // Extract timing information
        if (response_json.contains("total_duration")) {
            result.timings.ollama_total_ms = 
                response_json["total_duration"].get<int64_t>() / 1e6;
        }
        
        if (response_json.contains("load_duration")) {
            result.timings.ollama_load_ms = 
                response_json["load_duration"].get<int64_t>() / 1e6;
        }
        
        if (response_json.contains("prompt_eval_duration")) {
            result.timings.ollama_prompt_eval_ms = 
                response_json["prompt_eval_duration"].get<int64_t>() / 1e6;
        }
        
        if (response_json.contains("eval_duration")) {
            result.timings.ollama_eval_ms = 
                response_json["eval_duration"].get<int64_t>() / 1e6;
        }
        
        return true;
        
    } catch (const std::exception& e) {
        result.error = e.what();
        return false;
    }
}

VisionResult HybridVisionSystem::query(
    const std::string& prompt,
    const ImageData& image
) {
    VisionResult result;
    result.prompt = prompt;
    
    auto total_start = std::chrono::high_resolution_clock::now();
    
    std::cout << "\n" << std::string(70, '=') << std::endl;
    std::cout << "Querying Vision Model: " << config_.ollama_model << std::endl;
    std::cout << std::string(70, '=') << std::endl;
    std::cout << "Prompt: " << prompt << std::endl;
    std::cout << std::string(70, '-') << std::endl;
    
    if (!image.is_valid()) {
        result.error = "Invalid image data";
        return result;
    }
    
    // Step 1: ONNX preprocessing (optional)
    if (config_.use_onnx_preprocessing && has_onnx_preprocessing()) {
        std::cout << "\n1️⃣  ONNX Preprocessing..." << std::endl;
        
        std::vector<float> embeddings;
        float preprocess_time = preprocess_with_onnx(image, embeddings);
        
        if (preprocess_time > 0 && !embeddings.empty()) {
            result.timings.onnx_preprocess_ms = preprocess_time;
            result.embeddings = std::move(embeddings);
            
            std::cout << "   ✅ Image validated and preprocessed on GPU" << std::endl;
            std::cout << "   Time: " << preprocess_time << "ms" << std::endl;
        }
    }
    
    // Step 2: Optimize image for Ollama
    std::cout << "\n2️⃣  Optimizing image for Ollama..." << std::endl;
    
    std::string image_b64;
    float optimize_time = optimize_for_ollama(image, image_b64);
    
    result.timings.image_optimize_ms = optimize_time;
    
    std::cout << "   Original size: " << image.width << "x" << image.height << std::endl;
    std::cout << "   Encoded size: " << (image_b64.size() / 1024.0) << " KB" << std::endl;
    std::cout << "   Time: " << optimize_time << "ms" << std::endl;
    
    // Step 3: Query Ollama
    std::cout << "\n3️⃣  Sending to Ollama..." << std::endl;
    
    result.success = query_ollama(prompt, image_b64, result);
    
    if (result.success) {
        std::cout << "   ✅ Response received" << std::endl;
    } else {
        std::cout << "   ❌ Error: " << result.error << std::endl;
    }
    
    auto total_end = std::chrono::high_resolution_clock::now();
    result.timings.total_ms = 
        std::chrono::duration<float, std::milli>(total_end - total_start).count();
    
    return result;
}

std::optional<ImageData> HybridVisionSystem::load_image(const std::string& path) {
    ImageData image;
    
    int width, height, channels;
    uint8_t* data = stbi_load(path.c_str(), &width, &height, &channels, 3);
    
    if (!data) {
        std::cerr << "Failed to load image: " << path << std::endl;
        return std::nullopt;
    }
    
    image.width = width;
    image.height = height;
    image.channels = 3;  // Force RGB
    image.pixels.assign(data, data + (width * height * 3));
    
    stbi_image_free(data);
    
    return image;
}

std::optional<ImageData> HybridVisionSystem::load_image_from_memory(
    const std::vector<uint8_t>& data
) {
    ImageData image;
    
    int width, height, channels;
    uint8_t* pixels = stbi_load_from_memory(
        data.data(),
        static_cast<int>(data.size()),
        &width, &height, &channels, 3
    );
    
    if (!pixels) {
        std::cerr << "Failed to load image from memory" << std::endl;
        return std::nullopt;
    }
    
    image.width = width;
    image.height = height;
    image.channels = 3;
    image.pixels.assign(pixels, pixels + (width * height * 3));
    
    stbi_image_free(pixels);
    
    return image;
}

// ============================================================================
// VisionSystemManager Implementation
// ============================================================================

std::unique_ptr<HybridVisionSystem> VisionSystemManager::instance_ = nullptr;
bool VisionSystemManager::initialized_ = false;

HybridVisionSystem& VisionSystemManager::get_instance() {
    if (!instance_) {
        instance_ = std::make_unique<HybridVisionSystem>();
        initialized_ = true;
    }
    return *instance_;
}

bool VisionSystemManager::initialize(const VisionConfig& config) {
    try {
        instance_ = std::make_unique<HybridVisionSystem>(config);
        initialized_ = true;
        return true;
    } catch (const std::exception& e) {
        std::cerr << "Failed to initialize vision system: " << e.what() << std::endl;
        return false;
    }
}

void VisionSystemManager::shutdown() {
    instance_.reset();
    initialized_ = false;
}

} // namespace vision
} // namespace grim
